### PR TITLE
[FSI][Trilinos] FSI upgrade to analysis

### PR DESCRIPTION
**Pull request by rubenzorrilla**
Since this is a quite large PR I'd try to more or less split the review. Here it is my proposal:

@philbucher I've upgraded the FSI app. to work with the stage. Since you've worked a lot on that, would you mind to have a look to it and give me some feedback? (@RiccardoRossi I now that it is quite different to the base one, but as we discussed this is the way to do with the current interface. We can update it again in the future). @philbucher you will also notice that I took the chance to enable Cotire in the MappingApplication

@jcotela I've done some slightly modifications in the redistribution utility to make it MPI compatible (basically to add some calls to the communicator). Besides, I did modify the trilinos application because of the requirements of the FSI application. As you know, since PyBind the vector pointer manipulation from Python is a bit more complicated, so I did change the API of the FSI interface vectors utility. This required the definition of the Vector/MatrixPointerWrappers in a new header (trilinos_pointer_wrapper.h).  I'll also appreciate if you have a look to the Python solvers (nothing new, they follow the structure of the recent modifications in the fluid and structure ones).

@RiccardoRossi Please, have a look to the new wrapper (I think that this is quite important and I want to be sure that everything is fine). 

Thank you all!

### CODE REVIEW

File: applications/MappingApplication/CMakeLists.txt
Comment (philbucher) [inline applications/MappingApplication/CMakeLists.txt:42]: thanks

I finally have to try this ...

File: applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h
Comment (philbucher) [inline applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h]: Can you add the comment why this class is needed?
Basically what you wrote in the first post of the PR

File: applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_base_solver.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_base_solver.py]: side remark: this should be quite simple, you only need the `ADD_VALUES` argument for the mapping functions of the 2-n mapper to not overwrite the previous values

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (philbucher) [inline applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: KRATOS_WARNING_IF

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (jcotela) [inline applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: This is problematic. The local model part might not have conditions, in which case this call will fail. We probably need to synchronize the values of GeometryFamily and PointsNumber to be safe.

File: applications/FSIapplication/python_scripts/partitioned_fsi_base_solver.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/partitioned_fsi_base_solver.py]: I would compare agains an epsilon, these are doubles

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (jcotela) [inline applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: Same as above, one given partition might not have conditions and these calls will fail.

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (jcotela) [inline applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: I'd do a sum, to be equivalent with the serial version.

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/fsi_analysis.py:64]: The restart is now handled differently => such that not every stage has to copy paste the code from the other ones

I sent @RiccardoRossi a working example yesterday, it is much simpler

Then I think you could use `Initilaize` from the baseclass

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/fsi_analysis.py:17]: why don't you call the baseclass constructor?

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/fsi_analysis.py:99]: Instead of copy pasting lines 99-105 I would call the baseclass-InitializeSolutionStep

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/fsi_analysis.py:107]: If you use the new way for the output (and the restart) then you don't need this function
=> have a look at the structure or the fluid stage

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/fsi_analysis.py:148]: This function seems to hardcode the "old" way of defining the processes (=> without the encapsulation in "processes")
This is a bit limiting in my eyes but your call ofc

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/fsi_analysis.py:173]: If you add this to the list of output_processes then all the calls are taken care of corectly in the baseclass (PrintOutput & IsOutputStep)

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/fsi_analysis.py:178]: I feel like you could simplify your life a lot using the stuff in the baseclass

**Review by philbucher — COMMENTED** [review]
Some comments for improvements. You mention that some things are done differently than in the baseclass, which is also what I marked. You could simplify your life  by using those but I guess you have your reasons.
If sth is missing can you let us know, then we might have to enhance the baseclass

Esp with the restart you can simplify your life a lot!
Have a look at the example I sent to @RiccardoRossi 
I will also send it to you

@jcotela @rubenzorrilla I think the redistibution utility is sth we could move to the core, it is very general I'd say

@rubenzorrilla could you please also add a test for the fsi-stage? I really feel bad each time I break sth :/

File: applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h
Comment (jcotela) [inline applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h]: Probably does not need to be included.

File: applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h
Comment (jcotela) [inline applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h]: I believe this should be `typename TrilinosSparseSpaceType::MatrixPointerType` (gcc is more flexible with this than the standard, other compilers will not compile this).

Anyway, since you are using this in quite a few places, I'd suggest a `typedef` or `using` statement.

**Review by jcotela — CHANGES_REQUESTED** [review]
Overall good work, but I think there are some potential distributed memory problems, maybe we can discuss how to solve them in a call.

Also, I have not reviewed the analysis class in detail, but I tend to agree with @philbucher that some parts could just be inherited from the base class by following the "standard" more closely.

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (philbucher) [reply applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: I think the solution would be to check both for local and global num_conds

Comment (jcotela) [reply applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: @philbucher This could work, but then we cannot perform collective MPI operations within the function call (because the process that has no local conditions will never enter it).

Comment (rubenzorrilla) [reply applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: Totally agree, I did came across with this problem during this morning... I'm currently working on it.

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (rubenzorrilla) [reply applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: Done.

File: applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h
Comment (rubenzorrilla) [reply applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h]: I'm compiling with clang, without any problem. In any case, I'll define a `typedef`

File: applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h
Comment (rubenzorrilla) [reply applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h]: Removed!

File: applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h
Comment (jcotela) [reply applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h]: Maybe I'm wrong then... Try it and if it is not necessary it should not compile.

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/fsi_analysis.py:64]: I will have a look. Thanks @philbucher

File: applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h
Comment (rubenzorrilla) [reply applications/trilinos_application/custom_python/trilinos_pointer_wrapper.h]: Done.

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (jcotela) [inline applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: I'm afraid you also need a dummy here, since `SpecializedConvertDistributedValuesToPoint` has a Synchronize inside.

**Review by jcotela — CHANGES_REQUESTED** [review]

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (rubenzorrilla) [reply applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: In principle this is fixed with the new "dummy" methods

File: applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp
Comment (rubenzorrilla) [reply applications/FSIapplication/custom_utilities/variable_redistribution_utility.cpp]: Actually, we need to do everything as in the `CallSpecializedDistributePointValues` method. Thanks for noticing!

File: applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_base_solver.py
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_base_solver.py]: Done! Thanks @philbucher !

File: applications/FSIapplication/python_scripts/partitioned_fsi_base_solver.py
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_fsi_base_solver.py]: Done.

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/fsi_analysis.py:17]: Because the current structure of the FSI application ProjectParameters.json is not compatible with it. The point is that we need to change the GiD interface writing script for such FSI app, Since this is not a minor task, we decided to keep it as is done in this PR as a temporary solution,

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/fsi_analysis.py:99]: It is not possible, since the baseclass-InitializeSolutionStep calls the GetComputingModelPart() of the solver. Since in the partitioned FSI solver there are two "computing model parts" (fluid and structure) this method is not implemented.

**Comment by rubenzorrilla** [comment]
@philbucher I think that I will wait to implement your suggestions (even though we are in agreement) until we update the GiD interface. We will try to fit as much as possible into the core analysis stage.
@jcotela In principle, the variable redistribution utility should MPI compatible with the lasts modifications.

File: applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_dirichlet_neumann_solver.py
Comment (philbucher) [inline applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_dirichlet_neumann_solver.py]: this is not working for two reasons:
the correct python syntax is:
`KratosMapping.Mapper.SWAP_SIGN | KratosMapping.Mapper.ADD_VALUES)`

Furthermore you should not specify `ADD_VALUES` for the first mapper, only for the following ones
The first one should actually overwrite the previous values

File: applications/FSIapplication/python_scripts/fsi_analysis.py
Comment (philbucher) [reply applications/FSIapplication/python_scripts/fsi_analysis.py:99]: I see
I also don't like the current solution, we should come up with a better one for this @RiccardoRossi @jcotela 
=> I think we only need the time/step from the solver, which could be implemented in a separate function
I will think abt it

**Review by philbucher — DISMISSED** [review]
The trilinos mapping will not work in the current configuration

File: applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_dirichlet_neumann_solver.py
Comment (philbucher) [reply applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_dirichlet_neumann_solver.py]: you can take a look at the [wiki](https://github.com/KratosMultiphysics/Kratos/wiki/How-to-do-Mapping-between-nonmatching-meshes#using-the-mapping-functions)

Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/trilinos_partitioned_fsi_dirichlet_neumann_solver.py]: Done. Thanks for noticing!

**Review by jcotela — APPROVED** [review]
Approving with regards to the variable redistribution utility.

**Pull request merged**
